src/mono/mono/mini/interp/transform.c

```diff
@@ -7374,6 +7437,23 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 			interp_add_ins (td, MINT_NOP);
 	}
 
+	// Mark the last block with emitting information (if not already marked)
+	// and retry code generation if there was any skipped block.
+	td->cbb->already_emitted |= emitting;
```

Comment:
Why is this different from earlier:

```c
if (new_bb != NULL && td->cbb != new_bb) {
			// Mark the current block with emitting information
			td->cbb->already_emitted = emitting;
```

Reply:
The `|=` on line:7442 is required for the last blocks in order to avoid the following problem:
- in the first pass the last block is marked as emitted
- in the second pass a block is skipped (`emitting` flag set to `FALSE`)
- all the adjacent blocks have been emitted and the loop exits due to `td->ip = end;`condition
- as the loop exits, the last block's `already_emitted` is overwritten with `FALSE`

The 'earlier' case in question is protected by the first condition in the loop where all current blocks are skipped which have the `already_emitted` set, so the overwriting can never happen.

As you mentioned in your previous comments the loop is really sensitive to several boolean flags which make it harder to maintain. I agree with your proposal of introducing state enums which would bring clarification for these kind of corner cases